File: app/src/main/java/com/limelight/binding/input/advance_setting/KeyboardGestureDetector.java
package com.limelight.binding.input.advance_setting;

import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;

public class KeyboardGestureDetector {

    public interface GestureListener {
        void onKeyPress(int keyCode);
        void onKeyRelease(int keyCode);
        void onModifierHoldRelease(int keyCode);
        void onLongPress(int keyCode);
        void onDoubleTap(int keyCode);
    }

    private final GestureListener listener;
    private final int touchSlop;
    private long lastDownTime = 0;
    private long downTimeMs = 0;
    private static final int DOUBLE_TAP_TIMEOUT = 250;
    private static final int HOLD_THRESHOLD = 200;

    public KeyboardGestureDetector(View view, GestureListener listener) {
        this.listener = listener;
        this.touchSlop = ViewConfiguration.get(view.getContext()).getScaledTouchSlop();
    }

    public boolean onTouchEvent(View v, MotionEvent event) {
        String tag = (String) v.getTag();
        if (tag == null || !tag.startsWith("k")) return false;
        int keyCode = Integer.parseInt(tag.substring(1));

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                v.setPressed(true);
                v.refreshDrawableState();
                downTimeMs = System.currentTimeMillis();
                
                long currentTime = System.currentTimeMillis();
                if (currentTime - lastDownTime < DOUBLE_TAP_TIMEOUT) {
                    listener.onDoubleTap(keyCode);
                    lastDownTime = 0; // Reset to prevent triple-taps triggering it again
                } else {
                    listener.onKeyPress(keyCode);
                    lastDownTime = currentTime;
                }
                return true;

            case MotionEvent.ACTION_MOVE:
                return true;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                v.setPressed(false);
                v.refreshDrawableState();
                long pressDuration = System.currentTimeMillis() - downTimeMs;
                if (pressDuration >= HOLD_THRESHOLD) {
                    listener.onModifierHoldRelease(keyCode);
                } else {
                    listener.onKeyRelease(keyCode);
                }
                return true;
        }
        return false;
    }
}

File: app/src/main/java/com/limelight/binding/input/advance_setting/KeyboardGestureDetector.kt
package com.limelight.binding.input.advance_setting

import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration

class KeyboardGestureDetector(view: View, private val listener: GestureListener) {

    interface GestureListener {
        fun onKeyPress(keyCode: Int)
        fun onKeyRelease(keyCode: Int)
        fun onModifierHoldRelease(keyCode: Int)
        fun onLongPress(keyCode: Int)
        fun onDoubleTap(keyCode: Int)
    }

    private val touchSlop = ViewConfiguration.get(view.context).scaledTouchSlop
    private var lastDownTime = 0L
    private var downTimeMs = 0L

    fun onTouchEvent(v: View, event: MotionEvent): Boolean {
        val tag = v.tag as? String ?: return false
        if (!tag.startsWith("k")) return false
        val keyCode = tag.substring(1).toInt()

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                v.isPressed = true
                v.refreshDrawableState()
                downTimeMs = System.currentTimeMillis()

                val currentTime = System.currentTimeMillis()
                if (currentTime - lastDownTime < DOUBLE_TAP_TIMEOUT) {
                    listener.onDoubleTap(keyCode)
                    lastDownTime = 0
                } else {
                    listener.onKeyPress(keyCode)
                    lastDownTime = currentTime
                }
                return true
            }

            MotionEvent.ACTION_MOVE -> return true

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                v.isPressed = false
                v.refreshDrawableState()
                val pressDuration = System.currentTimeMillis() - downTimeMs
                if (pressDuration >= HOLD_THRESHOLD) {
                    listener.onModifierHoldRelease(keyCode)
                } else {
                    listener.onKeyRelease(keyCode)
                }
                return true
            }
        }
        return false
    }

    companion object {
        private const val DOUBLE_TAP_TIMEOUT = 250
        private const val HOLD_THRESHOLD = 200
    }
}
